Add ` --flatten`, `--depth` and `flatten-exclude` flags to `pack builder create` and `pack buildpack package` command

jjbustamante opened the pull request:
## Summary

This PR adds new flags to:
 -  `pack builder create`
 -  `pack buildpack package`

The flags will try to flatten the layers in the final image.

for both commands the following flags were added:

```bash
      --depth int                   Max depth to flatten compose buildpacks layers.
                                    Omission of this flag or values < 0 will flatten the entire tree. (default -1)
      --flatten                     Flatten the composite buildpacks layers
  -e, --flatten-exclude strings     Buildpack to exclude for being flatten. Use buildpack by id and version in the form of '<buildpack>@<version>'
```

These flags can be used with the following combinations:
- `--flatten` only OR with `--depth` < 0: it will flatten **ALL** the buildpacks layers into 1 final layer, because the default value for `--depth` is < 0
- `--flatten` and `--depth` = 1: it will flatten only the top level of composites to be flattened (i.e. composites directly referenced from either the builder or directly referenced from a composite if building a buildpack)
- `--flatten` and `--depth` = 2: it will flatten the top 2 levels of composites buildpacks 

And so on 

In case the users want to **EXCLUDE** some buildpacks for being flatten, they can use the `--flatten-exclude` flag in combination with the previous one to exclude any buildpack they want. The buildpack *MUST* be specify using the format `flatten-exclude '<buildpack>@<version>'`

### Example

For testing the feature I used the buildpacks defined in our sample [repository](https://github.com/buildpacks/samples/tree/main/buildpacks) and a `builder.toml` with the following content:

```toml
# Buildpacks to include in builder
[[buildpacks]]
uri = "samples/buildpacks/hello-processes"

[[buildpacks]]
# Packaged buildpacks to include in builder;
# the "hello-universe" package contains the "hello-world" and "hello-moon" buildpacks
uri = "docker://cnbs/sample-package:hello-universe"

# Order used for detection
[[order]]
    # This buildpack will display build-time information (as a dependency)
    [[order.group]]
    id = "samples/hello-world"
    version = "0.0.1"

    # This buildpack will display build-time information (as a dependant)
    [[order.group]]
    id = "samples/hello-moon"
    version = "0.0.1"

    # This buildpack will create a process type "sys-info" to display runtime information
    [[order.group]]
    id = "samples/hello-processes"
    version = "0.0.1"

# Stack that will be used by the builder
[stack]
id = "io.buildpacks.samples.stacks.bionic"
# This image is used at runtime
run-image = "cnbs/sample-stack-run:bionic"
# This image is used at build-time
build-image = "cnbs/sample-stack-build:bionic"
```
Graphically we can imagine the example as follow

![Screenshot 2023-04-20 at 4 09 51 PM](https://user-images.githubusercontent.com/1181799/233488316-2028462a-5918-463a-8900-7d22aef8b208.png)

#### Output

```bash
$ pack builder create cnbs/sample-builder:alpine --config builder.toml --flatten
```

Here we can see all the buildpacks were flatten into one layer:

![Screenshot 2023-04-20 at 4 14 12 PM](https://user-images.githubusercontent.com/1181799/233489007-d4e3ac8f-5c5a-4abe-b2e1-6f423d6f563f.png)

```bash
$ pack builder create cnbs/sample-builder:alpine --config builder.toml --flatten --depth 0
```
It will flatten the `hello-universe` in one layer and 

![Screenshot 2023-04-20 at 4 16 14 PM](https://user-images.githubusercontent.com/1181799/233489359-8503e748-6ff8-406f-ac9d-67160bf230f4.png)

And the `hello-processes` in another layer

![Screenshot 2023-04-20 at 4 16 59 PM](https://user-images.githubusercontent.com/1181799/233489501-f54f0c35-37b7-404d-9652-29d7d6cd79cc.png)

But if we want to **EXCLUDE** we can do

```bash
$ pack builder create cnbs/sample-builder:alpine --config builder.toml --verbose --flatten --depth 0 --flatten-exclude samples/hello-world@0.0.1
```
![Screenshot 2023-04-20 at 4 20 01 PM](https://user-images.githubusercontent.com/1181799/233490150-9ece871e-bb6a-4ad7-80ed-517537af9237.png)

![Screenshot 2023-04-20 at 4 20 12 PM](https://user-images.githubusercontent.com/1181799/233490178-c6297d2c-ebbd-41de-bc67-7cd68e9b68b9.png)

![Screenshot 2023-04-20 at 4 20 40 PM](https://user-images.githubusercontent.com/1181799/233490328-5eefe721-6dfc-4ec8-83eb-47ae03be06de.png)

#### Metadata

Inspecting the metadata `io.buildpacks.buildpack.layers`

```bash
> docker inspect cnbs/sample-builder:alpine | jq '.[] | .Config.Labels | ."io.buildpacks.buildpack.layers" | fromjson'
```

 shows the `layerDiffID` value is the same for the squashed buildpacks

```json
{
  "samples/hello-extensions": {
    "0.0.1": {
      "api": "0.9",
      "stacks": [
        {
          "id": "*"
        }
      ],
      "layerDiffID": "sha256:baf4f7c28c15dbaadc953bb6d98ce9e83473e842fe92956ea2a328495237ddfc",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/hello-extensions",
      "name": "Hello Extensions Buildpack"
    }
  },
  "samples/hello-moon": {
    "0.0.1": {
      "api": "0.2",
      "stacks": [
        {
          "id": "io.buildpacks.samples.stacks.jammy"
        },
        {
          "id": "io.buildpacks.samples.stacks.alpine"
        },
        {
          "id": "io.buildpacks.stacks.jammy"
        },
        {
          "id": "*"
        }
      ],
      "layerDiffID": "sha256:4ad60d1ea1836600073b50ae066e946068fdba9babbee257d3584cbe1ceb9cb7",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/hello-moon",
      "name": "Hello Moon Buildpack"
    }
  },
  "samples/hello-universe": {
    "0.0.1": {
      "api": "0.2",
      "order": [
        {
          "group": [
            {
              "id": "samples/hello-world",
              "version": "0.0.1"
            },
            {
              "id": "samples/hello-moon",
              "version": "0.0.1"
            }
          ]
        }
      ],
      "layerDiffID": "sha256:4ad60d1ea1836600073b50ae066e946068fdba9babbee257d3584cbe1ceb9cb7",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/hello-universe",
      "name": "Hello Universe Buildpack"
    }
  },
  "samples/hello-world": {
    "0.0.1": {
      "api": "0.2",
      "stacks": [
        {
          "id": "io.buildpacks.samples.stacks.jammy"
        },
        {
          "id": "io.buildpacks.samples.stacks.alpine"
        },
        {
          "id": "io.buildpacks.stacks.jammy"
        },
        {
          "id": "*"
        }
      ],
      "layerDiffID": "sha256:4ad60d1ea1836600073b50ae066e946068fdba9babbee257d3584cbe1ceb9cb7",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/hello-world",
      "name": "Hello World Buildpack"
    }
  },
  "samples/java-maven": {
    "0.0.1": {
      "api": "0.2",
      "stacks": [
        {
          "id": "io.buildpacks.samples.stacks.jammy"
        },
        {
          "id": "io.buildpacks.samples.stacks.alpine"
        },
        {
          "id": "*"
        },
        {
          "id": "io.buildpacks.stacks.jammy",
          "mixins": [
            "build:set=shell-utils"
          ]
        }
      ],
      "layerDiffID": "sha256:a74e11d973b550235a0f9f1041bdeb0d810d20f5f646a53465d812a925c8b671",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/java-maven",
      "name": "Sample Java Maven Buildpack"
    }
  },
  "samples/kotlin-gradle": {
    "0.0.1": {
      "api": "0.2",
      "stacks": [
        {
          "id": "io.buildpacks.samples.stacks.jammy"
        },
        {
          "id": "io.buildpacks.samples.stacks.alpine"
        },
        {
          "id": "*"
        },
        {
          "id": "io.buildpacks.stacks.jammy",
          "mixins": [
            "build:set=shell-utils"
          ]
        }
      ],
      "layerDiffID": "sha256:5511b6ad86c688fc70ef7056a21f7dcf82eff299c9e03d05cfb8dc47c7e37b90",
      "homepage": "https://github.com/buildpacks/samples/tree/main/buildpacks/kotlin-gradle",
      "name": "Sample Kotlin Gradle Buildpack"
    }
  }
}

```

Note: `pack buildpack package has a similar behavior`

#### Before

When the `pack builder create` is executed without the flag, each buildpack is added into a different layer

## Documentation
<!-- If this change should be documented, please create an issue or PR on https://github.com/buildpacks/docs and link below. -->
<!-- NOTE: This can be added (by editing the issue) after the PR is opened. -->

- Should this change be documented?
    - [ ] Yes, see #___
    - [ ] No

## Related
<!-- If this PR addresses an issue, please provide issue number below. -->

Resolves #1595

jjbustamante commented:
I still need to work in the unit test coverage, and probably I will like to refactor a little bit more, but I think in general the PR is ready for review.

Another thing I need to take a look is adding the `depth` flag Natalie mentioned.

natalieparellano commented:
I want to bikeshed a little bit about the `--flatten-meta-buildpacks` flag name. I think technically according to our [terminology](https://github.com/buildpacks/spec/blob/main/buildpack.md#cnb-terminology) it should be `--flatten-composite-buildpacks`. If we do `--depth` we might only need `--flatten`.

jkutner commented:
I agree with @natalieparellano on `--flatten-composite-buildpacks`. But I didn't understand the `--depth` comment.

Also, is "flatten" common terminology for this? It feels more like "merge" or "combine". but if flatten is used in other places that's fine.

jjbustamante commented:
> I agree with @natalieparellano on `--flatten-composite-buildpacks`. But I didn't understand the `--depth` comment.
> 
> Also, is "flatten" common terminology for this? It feels more like "merge" or "combine". but if flatten is used in other places that's fine.

"flatten" is used by `crane` for example:

```bash
$ crane flatten --help
Flatten an image's layers into a single layer

Usage:
  crane flatten [flags]

Flags:
  -h, --help         help for flatten
  -t, --tag string   New tag to apply to flattened image. If not provided, push by digest to the original image repository.

Global Flags:
      --allow-nondistributable-artifacts   Allow pushing non-distributable (foreign) layers
      --insecure                           Allow image references to be fetched without TLS
      --platform platform                  Specifies the platform in the form os/arch[/variant][:osversion] (e.g. linux/amd64). (default all)
  -v, --verbose                            Enable debug logs

```

jkutner commented:
@jjbustamante do you mind if we hold this until after 0.30.0-pre2?

jjbustamante commented:
> @jjbustamante do you mind if we hold this until after 0.30.0-pre2?

No problem, I am going to draft the PR again, but I am checking Nataliee's feedback

dmikusa commented:
@jjbustamante I had a few mins to play around with this and things look good. I built your PR and ran it with one of the Paketo Java composite buildpacks. Things seemed to go OK. Layers were squashed as you described and what you've described is exactly what we're looking for in Paketo.

If it's not a lot more effort, I'd be 👍 for `--flatten` and `--depth`. I think the extra control could be helpful. If I understand the proposal for `--depth` right, I would then `--flatten` to flatten all, `--flatten --depth 1` if I wanted only the top level of composites to be flattened (i.e. composites directly referenced from either the builder or directly referenced from a composite if building a buildpack), `--flatten --depth 2` to flatten the top two levels, etc...

Having said that, I'm not sure we'd necessarily need that extra control on Paketo so if it's a lot of effort, then I think options to flatten-all and flatten top-level composites be just fine.

Thanks for working on this!

jjbustamante commented:
> @jjbustamante I had a few mins to play around with this and things look good. I built your PR and ran it with one of the Paketo Java composite buildpacks. Things seemed to go OK. Layers were squashed as you described and what you've described is exactly what we're looking for in Paketo.
> 
> If it's not a lot more effort, I'd be 👍 for `--flatten` and `--depth`. I think the extra control could be helpful. If I understand the proposal for `--depth` right, I would then `--flatten` to flatten all, `--flatten --depth 1` if I wanted only the top level of composites to be flattened (i.e. composites directly referenced from either the builder or directly referenced from a composite if building a buildpack), `--flatten --depth 2` to flatten the top two levels, etc...
> 
> Having said that, I'm not sure we'd necessarily need that extra control on Paketo so if it's a lot of effort, then I think options to flatten-all and flatten top-level composites be just fine.
> 
> Thanks for working on this!

Thanks for taking the time to test the branch @dmikusa ❤️ 

You are right about the expected behavior with the `--flatten` and `--depth`, that's what we are targeting. Another flag we are going to add is `--flatten-exclude` to give the user the opportunity to exclude any Buildpack they consider is better to have in a separate layer, maybe because they change frequently and avoid other Buildpacks to be retransferred.

jjbustamante commented:
> @jjbustamante I had a few mins to play around with this and things look good. I built your PR and ran it with one of the Paketo Java composite buildpacks. Things seemed to go OK. Layers were squashed as you described and what you've described is exactly what we're looking for in Paketo.
> 
> If it's not a lot more effort, I'd be 👍 for `--flatten` and `--depth`. I think the extra control could be helpful. If I understand the proposal for `--depth` right, I would then `--flatten` to flatten all, `--flatten --depth 1` if I wanted only the top level of composites to be flattened (i.e. composites directly referenced from either the builder or directly referenced from a composite if building a buildpack), `--flatten --depth 2` to flatten the top two levels, etc...
> 
> Having said that, I'm not sure we'd necessarily need that extra control on Paketo so if it's a lot of effort, then I think options to flatten-all and flatten top-level composites be just fine.
> 
> Thanks for working on this!

jjbustamante commented:
@robdimsdale Thanks for gave it a try to the feature, let me try your use cases, I was working on test coverage yesterday and changed some stuff and maybe I broke something. I will let you know

jjbustamante commented:
@robdimsdale I just pushed a fix for the error:

```bash
ERROR: saving image: no compatible stacks among provided buildpacks
```

I tested locally the `python` buildpack and I managed to `--flatten` all the layers, you can give a try. I am still checking the other error with the builder

jjbustamante commented:
@robdimsdale regarding the error

```bash
ERROR: failed to write image to the following tags: [jammy-base-builder-flatten: loading image "jammy-base-builder-flatten". first error: embedded daemon response: Error processing tar file(duplicates of file paths not supported): ]
```

I also saw the error locally saving the builder into the daemon, but, when I use the `publish` flag and save it into `gcr.io` the image is saved, so it seems to be an issue with the docker daemon that I need to investigate further

jjbustamante commented:
@robdimsdale, I think I fixed the error 

`Error processing tar file(duplicates of file paths not supported)`

I managed to create the builder for jammy, if you want, you can give it another try to the branch, I will appreciate that

jjbustamante commented:
@robdimsdale forget it, I am still having some performance issues. I will let you know

jjbustamante commented:
@robdimsdale I think I found the root cause of the issue: 
```bash
Error processing tar file(duplicates of file paths not supported)
```
The builder contains the same buildpack with different versions, like this one:

![Screenshot 2023-04-27 at 5 43 35 PM](https://user-images.githubusercontent.com/1181799/235007120-d55da547-dd68-40ba-bfc2-3e016da736eb.png)

Because of that, I think we ended up with some paths being duplicated, like: `/cnb/buildpacks/paketo-buildpacks_npm-install` and then docker is complaining about it

jjbustamante commented:
@robdimsdale , @dmikusa 

cc @natalieparellano 

Based on the [conversation](https://cloud-native.slack.com/archives/C0331B61A1Y/p1682638643609549) on slack, I managed to fix the error 

```bash
Error processing tar file(duplicates of file paths not supported)
```

I tested it with the [paketo jammy builder](https://github.com/paketo-buildpacks/builder-jammy-base) and the [paketo python buildpack](https://github.com/paketo-buildpacks/python). If you can give it another try, I will appreciate! Thanks Natalie for the help ❤️

natalieparellano commented on pkg/buildpack/buildpack_tar_writer_test.go:
```diff
@@ -44,8 +44,9 @@ func testBuildModuleWriter(t *testing.T, when spec.G, it spec.S) {
 
 	it.Before(func() {
-		logger = logging.NewLogWithWriters(&outBuf, &outBuf, logging.WithVerbose())
+		h.SkipIf(t, runtime.GOOS != "linux", "Skipped on non-linux")
```
What happens if we try to use this feature on Windows?

natalieparellano replied:
Uggh disregard, I thought I was reviewing the PR but it was only one commit 😞

jjbustamante replied:
I have to check all the Windows support, it's been hard to test the feature only in the ci runners, that's why I am skipping the tests there. But, I am trying to set up a VM to give it a try

natalieparellano commented on pkg/buildpack/module_manager_test.go:
```diff
@@ -0,0 +1,299 @@
+				})
+
+				it("returns false for no flatten modules", func() {
+					h.AssertFalse(t, moduleManager.IsFlatten(bp1))
+					h.AssertFalse(t, moduleManager.IsFlatten(compositeBP1))
```
I have to confess... I'm still not quite sure what `IsFlatten` is supposed to mean...

jjbustamante replied:
well, the idea is this: I will add Buildpacks to the *moduleManager* and the manager must take care of organizing those Buildpacks in two groups (according to the configuration provided): 
- Buildpacks that must be treated as individuals
- Buidlapcks that must be flattened

*IsFlatten* is supposed to answer the question: *Should this given Buildpack be treated as a flattened one?*

natalieparellano replied:
Ah... in this case, maybe a better name is `ShouldFlatten`?

natalieparellano commented on pkg/buildpack/builder_test.go:
```diff
@@ -712,6 +717,165 @@ func testPackageBuilder(t *testing.T, when spec.G, it spec.S) {
+				err          error
+			)
+			it.Before(func() {
+				h.SkipIf(t, runtime.GOOS == "windows", "Skipped on windows")
```
This **should** work on Windows so we should remove this skip if possible

natalieparellano reviewed (commented):
@jjbustamante this is really awesome to see. I'm still wrapping my head around some things, but here are some initial thoughts

natalieparellano commented on internal/builder/builder.go:
```diff
@@ -263,20 +288,60 @@ func (b *Builder) GID() int {
 	b.metadata.Buildpacks = append(b.metadata.Buildpacks, bp.Descriptor().Info())
 }
 
+func (b *Builder) AddBuildpacks(main buildpack.BuildModule, dependencies []buildpack.BuildModule) {
```
Is calling this method the same as ranging over an array of buildpacks and calling `b.AddBuildpack(bp)` on each?

natalieparellano commented on pkg/client/package_buildpack.go:
```diff
@@ -72,7 +81,12 @@ func (c *Client) PackageBuildpack(ctx context.Context, opts PackageBuildpackOpti
+	var packageBuilderOpts []buildpack.PackageBuilderOption
+	if opts.Flatten {
+		packageBuilderOpts = append(packageBuilderOpts, buildpack.WithFlatten(opts.Depth, opts.FlattenExclude),
+			buildpack.WithLayerWriterFactory(writerFactory), buildpack.WithLogger(c.logger))
```
What factory do we fall back to if this option is not provided?

jjbustamante replied:
The LayerWriterFactory is only used when you enable the flattened mode, so I didn't add a default value. Initially, I wanted to put all three things together: `WithFlatten(deps, exclude, factory, logging)` but then I separated on these 3 options.

natalieparellano replied:
I see... does `layerWriterFactory` even need to live on the `PackageBuilder` if we know that we always want to use a `*layer.WriterFactory`? Does adding this configuration option help testing?

jjbustamante commented on internal/builder/builder.go:
```diff
@@ -64,21 +64,22 @@ const (
+	layerWriterFactory       archive.TarWriterFactory
+	lifecycle                Lifecycle
+	lifecycleDescriptor      LifecycleDescriptor
+	additionalBuildpacks     buildpack.ModuleManager
```
Well, to me *Collection* will mean this is just a group on buildpacks, which is actually *true* but, it is more complex than that because it is a Collection of buildpacks organized according to some rules defined during the initialization of that Collection. That's why my first initial thought was to use the [Manager](https://www.eventhelix.com/design-patterns/manager/) or it could be a [facade](https://refactoring.guru/design-patterns/facade)

jjbustamante replied:
yeah, `ManagedCollection` looks go to me

jjbustamante replied:
Done

jjbustamante commented on internal/builder/builder.go:
```diff
@@ -263,20 +288,60 @@ func (b *Builder) GID() int {
 	b.metadata.Extensions = append(b.metadata.Extensions, bp.Descriptor().Info())
 }
 
+func (b *Builder) AddExtensions(main buildpack.BuildModule, dependencies []buildpack.BuildModule) {
```
Ohhh that's a good insight!! yeah, I agree then

jjbustamante commented on internal/builder/builder.go:
```diff
@@ -373,15 +438,22 @@ func (b *Builder) Save(logger logging.Logger, creatorMetadata CreatorMetadata) e
+		return err
+	}
+
+	err = b.addModules(buildpack.KindBuildpack, logger, tmpDir, b.image, append(b.additionalBuildpacks.NonFlattenModules(), excludedBuildpacks...), bpLayers)
```
I can't remember, let me check

jjbustamante replied:
Well, it is actually tested in all the existing tests, because it is returning the Buildpacks as they were before, but I do not think there is a explicit test for it

jjbustamante commented on pkg/client/create_builder_test.go:
```diff
@@ -975,6 +975,144 @@ func testCreateBuilder(t *testing.T, when spec.G, it spec.S) {
+					it("creates 1 layer for buildpacks and 1 layer for buildpack excluded", func() {
+						prepareFetcherWithRunImages()
+						opts.Flatten = true
+						opts.FlattenExclude = []string{"flatten/bp-7@7"}
```
Well, in this case, you still have 2 layers. Because during the exclusion I am not considering the hierarchy. so end users should exclude 1 by 1 if they want to exclude 5, 6, 7

jjbustamante commented on pkg/client/inspect_buildpack_test.go:
```diff
@@ -347,7 +348,10 @@ func testInspectBuildpack(t *testing.T, when spec.G, it spec.S) {
 	it.After(func() {
 		mockController.Finish()
-		h.AssertNil(t, os.RemoveAll(tmpDir))
+		err := os.RemoveAll(tmpDir)
```
Maybe, I think so.

jjbustamante replied:
I will add a TODO to keep track of this work

jjbustamante commented on pkg/client/package_buildpack_test.go:
```diff
@@ -413,6 +415,128 @@ func testPackageBuildpack(t *testing.T, when spec.G, it spec.S) {
+				})
+			})
+
+			when("with depth", func() {
```
Yeah, because you can also package a composite buildpack and maybe exclude some dependicies

natalieparellano replied:
We might want to add one test case for this

jjbustamante replied:
Also I will add a TODO to keep track of if

jjbustamante commented on internal/builder/builder_test.go:
```diff
@@ -1757,6 +1757,63 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+			builderImage = baseImage
+		})
+
+		when("flatten all", func() {
```
The only one could be "flatten with  depth", because "flatten none" is actually the other test cases that already exists which is the default behavior

natalieparellano replied:
Makes sense. What about something like:
```	when("flatten", func() {
		when("all", func() {
		})
		when("with depth", func() {
		})
	})
```

jjbustamante replied:
I will add a TODO, because the `with depth` test case requires to set a tree of composite buildpacks

jjbustamante commented on internal/builder/builder_test.go:
```diff
@@ -1757,6 +1757,63 @@ func testBuilder(t *testing.T, when spec.G, it spec.S) {
+				})
+			})
+
+			when("#AllModules", func() {
```
Will be like:
- `ModulesExploded()`: All the modules, default behavior which is an array of single BuildModules
- `ModulesFlattened()`: Modules organized by groups, each group and array, which will be converted into a layer
- `ModulesNonFlattened()`: When flattened, this are the individual modules that will not being flatten

natalieparellano replied:
I think I had the wrong understanding before. I would do:
* `AllModules()`: All the modules, default behavior which is an array of single BuildModules
* `FlattenedModules()`: Modules organized by groups, each group and array, which will be converted into a layer
* `ExplodedModules()`: When flattened, this are the individual modules that will not being flatten

I take back what I said about prefixing these methods with `Modules`, as it kind of implies 3 views of the same thing, whereas it's actually "all = flattened + exploded".

jjbustamante commented on pkg/buildpack/module_manager_test.go:
```diff
@@ -0,0 +1,299 @@
+				it("returns 1 flatten module with [compositeBP2, bp21, bp22, compositeBP3, bp31]", func() {
+					modules := moduleManager.FlattenModules()
+					h.AssertEq(t, len(modules), 1)
+					h.AssertEq(t, len(modules[0]), 5)
```
Yes, they are in `NonFlattenModules()`.

jjbustamante replied:
I added those test cases

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+type ModuleManager struct {
+	modules        []BuildModule
+	flattenModules [][]BuildModule
```
```suggestion
	flattenedModules [][]BuildModule
```

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+	return all
+}
+
+// NonFlattenModules returns all none flatten modules handle by the manager
+func (f *ModuleManager) NonFlattenModules() []BuildModule {
```
```suggestion
// ExplodedModules returns all modules that will be added to the output artifact as a single layer containing a single module.
func (f *Collection) ExplodedModules() []BuildModule {
```
Is that accurate? I think a slightly more descriptive comment would help others to understand it better.

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+	return f.modules
+}
+
+// FlattenModules returns all flatten modules handle by the manager.
+func (f *ModuleManager) FlattenModules() [][]BuildModule {
```
```suggestion
// FlattenedModules returns all modules that will be added to the output artifact as a single layer containing multiple modules.
func (f *Collection) FlattenedModules() [][]BuildModule {
```

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+	} else {
+		if f.maxDepth <= FlattenMaxDepth {
+			// flatten all
+			if len(f.flattenModules) == 1 {
+				f.flattenModules[0] = append(f.flattenModules[0], append([]BuildModule{main}, deps...)...)
+			} else {
+				f.flattenModules = append(f.flattenModules, append([]BuildModule{main}, deps...))
+			}
```
```suggestion
			f.flattenedModules = append(f.flattenedModules, append([]BuildModule{main}, deps...))
```
I think this will accomplish the same thing (please correct me if I'm wrong)

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+			}
+		} else {
+			recurser := newFlattenModuleRecurser(f.maxDepth)
+			calculateModules := recurser.calculateFlattenModules(main, deps, 0)
```
```suggestion
			calculatedModules := recurser.calculateFlattenModules(main, deps, 0)
```

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+	}
+}
+
+func (f *flattenModuleRecurser) calculateFlattenModules(main BuildModule, deps []BuildModule, depth int) [][]BuildModule {
```
Could we add a docs comment here?

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+
+func (f *flattenModuleRecurser) calculateFlattenModules(main BuildModule, deps []BuildModule, depth int) [][]BuildModule {
+	modules := make([][]BuildModule, 0)
+	orders := main.Descriptor().Order()
```
```suggestion
	groups := main.Descriptor().Order()
```
(there is only one order)

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+			modules = append(modules, append([]BuildModule{main}, deps...))
+		}
+		if depth < f.maxDepth {
+			bps, newDeps := buildpacksFromGroup(orders, deps)
```
```suggestion
			nextBPs, nextDeps := buildpacksFromGroup(groups, deps)
```
(this helped me understand it better)

natalieparellano commented on pkg/buildpack/module_manager.go:
```diff
@@ -0,0 +1,144 @@
+	return modules
+}
+
+func buildpacksFromGroup(orders dist.Order, deps []BuildModule) ([]BuildModule, []BuildModule) {
```
```suggestion
func buildpacksFromGroup(groups dist.Order, deps []BuildModule) ([]BuildModule, []BuildModule) {
```

natalieparellano replied:
(or you can use `order dist.Order`)

natalieparellano replied:
Could we add a docs comment here too?

natalieparellano commented on internal/builder/builder.go:
```diff
@@ -263,17 +288,49 @@ func (b *Builder) GID() int {
+	return buildpack.ModuleManager{}
+}
+
+func (b *Builder) FlattenModules(kind string) [][]buildpack.BuildModule {
```
```suggestion
func (b *Builder) FlattenedModules(kind string) [][]buildpack.BuildModule {
```

natalieparellano commented on internal/builder/builder.go:
```diff
@@ -263,17 +288,49 @@ func (b *Builder) GID() int {
+	return manager.FlattenModules()
+}
+
+func (b *Builder) MustBeFlatten(module buildpack.BuildModule) bool {
```
```suggestion
func (b *Builder) ShouldFlatten(module buildpack.BuildModule) bool {
```

natalieparellano commented on internal/builder/builder.go:
```diff
@@ -588,6 +660,73 @@ func (b *Builder) addModules(kind string, logger logging.Logger, tmpDir string,
 	return nil
 }
 
+func (b *Builder) addFlattenModules(kind string, logger logging.Logger, tmpDir string, image imgutil.Image, flattenModules [][]buildpack.BuildModule, layers dist.ModuleLayers) ([]buildpack.BuildModule, error) {
```
```suggestion
func (b *Builder) addFlattenedModules(kind string, logger logging.Logger, tmpDir string, image imgutil.Image, flattenModules [][]buildpack.BuildModule, layers dist.ModuleLayers) ([]buildpack.BuildModule, error) {
```

natalieparellano commented on internal/builder/builder.go:
```diff
@@ -393,10 +457,18 @@ func (b *Builder) Save(logger logging.Logger, creatorMetadata CreatorMetadata) e
+		return err
+	}
+
+	err = b.addModules(buildpack.KindExtension, logger, tmpDir, b.image, append(b.additionalExtensions.NonFlattenModules(), excludedExtensions...), extLayers)
```
```suggestion
	err = b.addExplodedModules(buildpack.KindExtension, logger, tmpDir, b.image, append(b.additionalExtensions.NonFlattenModules(), excludedExtensions...), extLayers)
```
(or something, whatever is consistent)

natalieparellano commented on pkg/buildpack/builder.go:
```diff
@@ -83,7 +142,27 @@ func (b *PackageBuilder) SetExtension(extension BuildModule) {
+	b.dependencies.AddModules(main, dependencies...)
+}
+
+func (b *PackageBuilder) MustBeFlatten(module BuildModule) bool {
```
```suggestion
func (b *PackageBuilder) ShouldFlatten(module BuildModule) bool {
```

natalieparellano commented on pkg/buildpack/builder.go:
```diff
@@ -83,7 +142,27 @@ func (b *PackageBuilder) SetExtension(extension BuildModule) {
+	return b.flattenAllBuildpacks || (b.dependencies.ShouldFlatten(module))
+}
+
+func (b *PackageBuilder) FlattenModules() [][]BuildModule {
```
```suggestion
func (b *PackageBuilder) FlattenedModules() [][]BuildModule {
```

natalieparellano commented:
@jjbustamante amazing work - the code seems well tested and appears to do what it sets out to do :) I added a few suggestions to hopefully improve the readability a bit, but the only that are truly blocking are:
* https://github.com/buildpacks/pack/pull/1691#discussion_r1185353617
* https://github.com/buildpacks/pack/pull/1691#discussion_r1187794019

jjbustamante commented on internal/builder/builder.go:
```diff
@@ -263,17 +288,49 @@ func (b *Builder) GID() int {
 }
 
+func (b *Builder) AllModules(kind string) []buildpack.BuildModule {
+	manager := b.moduleManager(kind)
+	return manager.AllModules()
```
I got `Cannot call a pointer method on 'b.moduleManager(kind)'`

natalieparellano replied:
Yuck, I think you can just make `func (b *Builder) moduleManager(kind string)` return `*buildpack.ModuleManager`

jjbustamante commented:
> @jjbustamante amazing work - the code seems well tested and appears to do what it sets out to do :) I added a few suggestions to hopefully improve the readability a bit, but the only that are truly blocking are:
> 
> * [Add ` --flatten`, `--depth` and `flatten-exclude` flags to `pack builder create` and `pack buildpack package` command #1691 (comment)](https://github.com/buildpacks/pack/pull/1691#discussion_r1185353617)
> * [Add ` --flatten`, `--depth` and `flatten-exclude` flags to `pack builder create` and `pack buildpack package` command #1691 (comment)](https://github.com/buildpacks/pack/pull/1691#discussion_r1187794019)

Thanks for this amazing feedback @natalieparellano ❤️ .. I just added some of them, I will take a look on the Windows support and the remaining suggestions! ..

jjbustamante commented:
> I'm reviewing this today

Thanks @jkutner, I just added some *TODO* to add more test coverage as @natalieparellano suggested, but the important things for me to fix are the *Windows* support and the improvement in the `flatten-exclude` flag to avoid compressing a composite buildpack and its dependencies without having to explicitly add the dependencies with the flag.

As we talk in the leadership meeting, I like the strategy for merging this PR to include it in the next pre-release, and I will create a new PR with the improvements mentioned above.

jkutner commented on pkg/client/testdata/buildpack-flatten/buildpack-7/buildpack.toml:
```diff
@@ -0,0 +1,9 @@
+api = "0.3"
```
are you using buildpack api 0.3 on purpose in these?

jjbustamante replied:
Not really. I think I copied and paste from another similar file
